### PR TITLE
dist: Add debian packaging

### PR DESCRIPTION
**Pull request by phlax**
Commit Message: dist: Add debian packaging
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: distribution/debian/packages.bzl
Comment (phlax) [inline distribution/debian/packages.bzl:98]: add comment explaining what this does

File: distribution/debian/preinst
Comment (phlax) [inline distribution/debian/preinst:33]: Improve installation message

File: distribution/distros.yaml
Comment (phlax) [inline distribution/distros.yaml:2]: add more debian distros

File: distribution/packages.bzl
Comment (phlax) [inline distribution/packages.bzl:20]: factor this out to a macro

File: distribution/packages.bzl
Comment (phlax) [inline distribution/packages.bzl:44]: or dont sign if they are not to be published / key doesnt exist

File: distribution/BUILD
Comment (phlax) [inline distribution/BUILD:19]: use a py_script with args for this

File: ci/do_ci.sh
Comment (phlax) [inline ci/do_ci.sh:222]: fix this

File: .azure-pipelines/pipelines.yml
Comment (phlax) [inline .azure-pipelines/pipelines.yml:334]: remove

**Comment by phlax** [comment]
@mklein @lizan @htuch

This PR adds the building and verification of deb packages to CI

In subsequent PRs im intending to:

- Add RPM support
- Add publishing/sync with netlify for release branches
- Figure out a publishing pipeline for historical versions

There is an ugly hack here to access the debian `.changes` files, which is non-hermetic/non-rbe-friendly.

I have raised an upstream bug for this (https://github.com/bazelbuild/rules_pkg/issues/477), so hopefully it will be resolved fairly soon.

There were a couple of blockers with RPMs, which i opened upstream PRs/issues for. I think these are now resolved so we should be able to add RPM support quite quickly.

Everything is setup on the netlify side for publishing, but we will need to figure out the release/publish/trigger workflow on the Envoy side to trigger this.

I have put the package and verify stages for arm/x86 into its own "distribution" stage. Im thinking this could/should probably be merged with the docker build stage, but not sure where is best to add these jobs to the CI

File: ci/do_ci.sh
Comment (phlax) [inline ci/do_ci.sh:222]: for this im wondering about moving it into the `envoy.gpg.sign` python lib - we can add a flag `--add-key` for non-release CI

File: distribution/debian/preinst
Comment (phlax) [inline distribution/debian/preinst:33]: move this to postinst

**Comment by wrowe** [comment]
I trust this PR remains draft for the moment until the wrinkles are ironed out?

**Comment by phlax** [comment]
> I trust this PR remains draft for the moment until the wrinkles are ironed out?

it should be ready for first review

**Comment by phlax** [comment]
not sure why the arm build is failing, it was working ok before

either way, it would be good to get this going through review

**Comment by phlax** [comment]
cc @lizan

File: distribution/debian/packages.bzl
Comment (lizan) [inline distribution/debian/packages.bzl:81]: where is .changes files generated/included and do we really need them? I'd rather remove this hack and go without .changes.

Comment (phlax) [reply distribution/debian/packages.bzl:81]: changes files are required to sign debs - i think its really not something we can do without

its generated by the `pkg_deb` rule and previously worked ~as expected - im hoping it will be resolved but not sure how long that will take (its been broken for 3-4 months)

**Comment by wrowe** [comment]
> not sure why the arm build is failing

Probably a timeout. Which means the changeset merging main had too many objects to recompile in a 2 hour window.

/retest

**Comment by wrowe** [comment]
I hope you won't mind since you've adopted this, trying to slot all the PR's to owners, thanks lizan.

/assign lizan

**Comment by wrowe** [comment]
Now give distribution arm64 another 2 hours ... I hope this is an incremental cached set of bazel actions.

/retest

**Comment by phlax** [comment]
ill take a look and see if i can figure why it has started failing - it was pretty reliable before - not sure if its because i moved it out of the "linux" job

File: ci/do_ci.sh
Comment (phlax) [reply ci/do_ci.sh:222]: i have implemented the required bits to move this into `envoy.gpg.sign` but when i tried running this under `bazel build` it fails to create the key

i have raised an upstream (`python-gnupg`) bug here https://bitbucket.org/vinay.sajip/python-gnupg/issues/161/gen_key-fails-when-run-under-bazel-build

File: distribution/debian/packages.bzl
Comment (phlax) [reply distribution/debian/packages.bzl:81]: seems like a fix for this is imminent https://github.com/bazelbuild/rules_pkg/pull/484

File: ci/do_ci.sh
Comment (phlax) [reply ci/do_ci.sh:227]: resolved

**Comment by wrowe** [comment]
Still...

/lgtm deps

Reason for the hangup was too many new cache objects to construct on 120 min, arm is slower than amd/Intel on these.

File: distribution/debian/packages.bzl
Comment (phlax) [reply distribution/debian/packages.bzl]: the hack for this has been removed, but currently it means using the above PR

i guess we will need to wait for it to land, but at least the hack is gone

File: distribution/debian/postinst
Comment (phlax) [inline distribution/debian/postinst:16]: add info about running as envoy user

File: distribution/debian/packages.bzl
Comment (phlax) [inline distribution/debian/packages.bzl:13]: add a not about ensuring these match `distros.yaml`

File: distribution/distros.yaml
Comment (phlax) [inline distribution/distros.yaml:1]: add a not about ensuring these match package `supported_distributions`

**Comment by jmarantz** [comment]
Can you merge main? Thanks!

/wait

**Comment by phlax** [comment]
> Can you merge main? Thanks!

done.

this PR is still waiting on https://github.com/bazelbuild/rules_pkg/pull/484

**Comment by alyssawilk** [comment]
I'll tag it as waiting then to get it out of the stale queue
/wait

File: distribution/debian/packages.bzl
Comment (MarkusTeufelberger) [inline distribution/debian/packages.bzl:13]:
```suggestion
        supported_distributions = "buster bullseye bionic focal hirsute impish",
```

Comment (MarkusTeufelberger) [reply distribution/debian/packages.bzl:13]: Though actually hirsute is already EOL: https://endoflife.date/ubuntu

Comment (phlax) [reply distribution/debian/packages.bzl:13]: hi, thanks for headsup ill remove it

**Comment by MarkusTeufelberger** [comment]
The PR in rules_pkg has been merged (though it was documentation only?). Anything else that's missing or that you need help with?

**Comment by phlax** [comment]
> The PR in rules_pkg has been merged (though it was documentation only?). Anything else that's missing or that you need help with?

not immediately, altho there are a few things

 i saw the PR was merged and pulled the current PR for `rules_pkg` here #19677 - we can update that to the current main HEAD pending a release - feel free to PR

the other small blocker is that the libs for building/testing have been refactored recently, and as they were not yet being used pending the rules_pkg update, i didnt update them - i will do that today/tomorrow - this is why CI is currently failing

once they are updated we should be good to go on adding this pipeline, and then we can look at adding rpms, and how to set up publishing for historical versions

there is also a little work to be done re the actual hosting/publishing on netlify, but not massive

**Comment by phlax** [comment]
are you on slack @MarkusTeufelberger ?

**Comment by phlax** [comment]
> though it was documentation only?

ah, i missed that, i think originally that PR did change something - but i realize now the fix that it suggests is already available - ie no pressing need to update the rules_pkg version beyond my current PR - ill reopen...

**Comment by MarkusTeufelberger** [comment]
> are you on slack @MarkusTeufelberger ?

Not if I can avoid it. :-/
Signed in for now again, but in general I really dislike that platform.

**Comment by phlax** [comment]
> Not if I can avoid it. :-/

8/

>  Signed in for now again, but in general I really dislike that platform.

well, without prejudice 8/, it is where the Envoy community is so im on there - there is a low traffic #envoy-dist channel which is probs a good place for general chat on distribution/packaging strategy etc

**Comment by MarkusTeufelberger** [comment]
Well, https://envoyslack.cncf.io/ seems broken anyways, so there is no way to join (I thought you were using the actual CNCF slack, not yet another one).

Aaaanyways, thanks for the update! I just spend quite a while on building current Debian packages for `envoy` with Bazel since the ones by tetrate linked in the documentation are out of date, unmaintained and still the only place to get somewhat "official" looking packaged binaries. There might be some general improvements possible/needed(?) beyond what's in this PR, but that's probably better discussed and implemented after this lands as an initial MVP.

**Comment by phlax** [comment]
yep, apologies that this has not been functional for a while, im working on making sure we have a properly tested/validated publising pipeline moving forward

i would appreciate any help/advice/Prs/etc so please do contribute to this effort further if you are able to

thanks for the headsup about the cncf slack page - ill raise this

in the meantime i think  you can access here https://envoy.slack.com/

**Comment by phlax** [comment]
@lizan i have a fix for the failing CI on this PR and will merge main and update as soon as it is published

im thinking to separate out the bazel <> pipeline changes and to land the bazel changes first - that way it will be a lot easier to see what is being added to ci and where, and make it easier to review the bazel part i think

lmk if you are happy with this plan, and ill update the PR accordingly

**Comment by MarkusTeufelberger** [comment]
Sure, just having a starting point to add onto is already great. It might actually be beneficial to not start publishing packages right away to get a few iterations in o the existing ones.

**Comment by phlax** [comment]
this should hopefully work now, ill update shortly...

**Comment by phlax** [comment]
@lizan this appears to be passing CI (apart from unrelated verify fail) and no longer has the hack etc. 

It should be ready for review

What do you think about my suggestion to separate landing the bazel parts from actual CI/pipeline hookup

Im not totally happy with where things are placed now and i think it will be easier to review and give proper consideration to where we put this in CI, if we separate to different PRs

**Comment by phlax** [comment]
ill try merging main - not sure what the issue is

**Comment by rojkov** [comment]
The tests are all green now, but we've got a merge conflict.

/wait

**Comment by phlax** [comment]
ill update again later today

**Comment by lizan** [comment]
@phlax CI is still failing?

**Comment by lizan** [comment]
@phlax CI is still failing?

**Comment by phlax** [comment]
its more like failing anew

not sure exactly why this problem is only emerging now - but when i was testing with the new python3.10 build image i was hitting a similar problem

in this case its only happening on arm

the solution i think is what was added here...

https://github.com/envoyproxy/envoy-build-tools/blob/f3a70cf47bd91d8f6ab080e4da361cc7f8e6f24d/build_container/build_container_ubuntu.sh#L163-L164

im just waiting on failing windows build of the build image and then i will update the image here in envoy - im guessing that will fix

**Comment by phlax** [comment]
it seems that may not have fixed it - not sure why - but i will debug further

if absolutely necessary, we can forego the `psutil` dep (which is struggling to build on arm) as its only used for debugging - but i would rather get it working if its possible

**Comment by phlax** [comment]
psutil has been removed as it wasnt really necessaary - so hopefully tests will be passing now

**Comment by MarkusTeufelberger** [comment]
The error seems to be here:

https://dev.azure.com/cncf/envoy/_build/results?buildId=103545&view=logs&j=f59d5e5d-007f-53ca-e54f-83f694f140b3&t=98a46940-48cd-528d-236d-233ea940bed2&l=153

`error: command 'aarch64-linux-gnu-gcc' failed: No such file or directory`

When building psutil 5.9.0 on ARM, gcc is not installed/available during the build. for some reason compilation is done via clang, but linking via `gcc` and this fails. https://hub.docker.com/r/envoyproxy/envoy-build/tags does not seem to contain the actual image used when building but https://github.com/envoyproxy/envoy-build-tools/blob/f3a70cf47bd91d8f6ab080e4da361cc7f8e6f24d/build_container/README.md refers there... :-/

I looked further and found https://console.cloud.google.com/gcr/images/envoy-ci/global/envoy-build and https://hub.docker.com/r/envoyproxy/envoy-build-ubuntu/tags which sounds more up-to-date.

At the moment I don't have an ARM64 machine at hand, but it might be worth checking out why the digest for that image is `Digest: sha256:4010a8310fc96227b9927f9ea3aa5c12913ef457d72be61099cc0f6eaf26a413` in CI and not the one on dockerhub (https://hub.docker.com/layers/envoyproxy/envoy-build-ubuntu/d859a503314ae611bb7ca4a7b4b4a19194e199f0/images/sha256-170b0f7cc4e157ca426c078c871da7cef9b59cab8c2166432a464ba5312d1efa?context=explore)?

**Comment by phlax** [comment]
@MarkusTeufelberger the problem is known and now fixed upstream (building psutil on arm)

**Comment by phlax** [comment]
(see https://github.com/envoyproxy/pytooling/commit/abb6e3cb41ec568984e4924e2eeb0d9292b3c676 for more info)

**Comment by MarkusTeufelberger** [comment]
I guess that's one way to fix this. ;-)

Might still be worth checking out if there is an actual `aarch64-linux-gnu-gcc` available in your build container in case some other library of yours actually needs it.

**Comment by phlax** [comment]
> Might still be worth checking out if there is an actual aarch64-linux-gnu-gcc available in your build container in case some other library of yours actually needs it.

re this - see this commit - https://github.com/envoyproxy/envoy-build-tools/pull/166/files#diff-de88d82423c82fcf8e66cf5d50f69b4de3ca587502583ffa1fdd0c48b504cadeR162-R163 - altho im a bit baffled why its not working

atm, rules_python doesnt have good support for cython on mac - and non-precompiled-but-needs-to-be packages on arm are best avoided - even if they can be made to work they often add a lot of build time

that said - it would be good to know why this hasnt worked - altho wrt to the psutil dep, the case for keeping it was too weak to justify spending any time on it

re arm on docker - i think qemu is probs the easiest way to test on an x86 machine (see eg https://www.stereolabs.com/docs/docker/building-arm-container-on-x86/)

**Comment by MarkusTeufelberger** [comment]
I think I found the issue with your build container build:

https://github.com/envoyproxy/envoy-build-tools/blob/f3a70cf47bd91d8f6ab080e4da361cc7f8e6f24d/build_container/build_container_common.sh#L80

**Comment by phlax** [comment]
maybe - altho im wondering whether any of the subsequent code there would be necessary - gcc is already installed at this point, and links etc should have been made - defo a possibility tho

**Comment by phlax** [comment]
aargghh - this is still an issue as `yarl` has to be compiled on 3.10 (at least for ARM)

re general points on the image @MarkusTeufelberger - i would encourage you to raise ticket/s in the envoy-build-tools repo and/or raise it on the relevant slack channel

**Comment by phlax** [comment]
@MarkusTeufelberger I have raised a PR to update the build image with your fix here #20391 

Ive also picked the update on this PR, so hopefully the arm build should be fixed now - thanks for that

**Comment by phlax** [comment]
cc @lizan tests are passing again

File: ci/do_ci.sh
Comment (daixiang0) [inline ci/do_ci.sh:242]: Is it real mail? I do not see it in the doc.

Comment (phlax) [reply ci/do_ci.sh:242]: it is not - and is intended to denote that - perhaps this could be more explicit in some way

**Comment by phlax** [comment]
@lizan would you mind giving this another pass - im v open to any suggestions on how we structure this

**Comment by adisuissa** [comment]
@lizan can you PTAL?

**Comment by alyssawilk** [comment]
@lizan this has been stalled 9 days waiting on your review.  Can you please take a look today?

**Comment by phlax** [comment]
ill merge main now, and check its not failing

File: ci/do_ci.sh
Comment (wrowe) [reply ci/do_ci.sh:242]: Use envoy@packaging.example.com perhaps?

Comment (phlax) [reply ci/do_ci.sh]: updated to envoy-ci@for.testing.only

**Comment by lizan** [comment]
Check CI? Otherwise LGTM.
/retest

**Review by zuercher — APPROVED** [review]
Approving based on @lizan's comment now that the tests passed. @phlax I'll let you merge it when you're ready.

**Comment by phlax** [comment]
ill merge this now - i suspect this will take a few iterations to optimize, but this is a major milestone/component to adding a release pipeline

**Pull request merged**
